rust/src/push/utils.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::bail;
+use anyhow::Context;
+use anyhow::Error;
+use lazy_static::lazy_static;
+use regex;
+use regex::Regex;
+use regex::RegexBuilder;
+
+lazy_static! {
+    /// Matches runs of "glob" style wild cards
+    static ref WILDCARD_RUN: Regex = Regex::new(r"([^\?\*]*)([\?\*]*)").expect("valid regex");
```

Comment:
docstring a little bit inaccurate; to my reading this matches the non-wildcards as group 1 followed by the wildcards as group 2.
Does it concern us that this regex can match an empty string? I suspect not if you've tried this

Reply:
Matching the empty string is fine, it'll cause us to compile a regex of an empty string (which is what we want). Plus in practice the usage of this is always gated behind checking if the glob has any wildcards in.